File: src/core/data_formats/parsing/quest/QuestObject.ts
import { id_to_object_type, object_data, ObjectType } from "./object_types";
import { Vec3 } from "../../vector";
import { OBJECT_BYTE_SIZE } from "./dat";
import { assert } from "../../../util";
import { angle_to_rad, rad_to_angle } from "../ninja/angle";
import { set_object_default_data } from "./set_object_default_data";

export type QuestObject = {
    area_id: number;
    readonly data: ArrayBuffer;
    readonly view: DataView;
};

export function create_quest_object(type: ObjectType, area_id: number): QuestObject {
    const data = new ArrayBuffer(OBJECT_BYTE_SIZE);
    const obj: QuestObject = {
        area_id,
        data,
        view: new DataView(data),
    };

    set_object_default_data(type, obj.view);

    set_object_type(obj, type);

    return obj;
}

export function data_to_quest_object(area_id: number, data: ArrayBuffer): QuestObject {
    assert(
        data.byteLength === OBJECT_BYTE_SIZE,
        () => `Data byteLength should be ${OBJECT_BYTE_SIZE} but was ${data.byteLength}.`,
    );

    return {
        area_id,
        data,
        view: new DataView(data),
    };
}

//
// Simple properties that directly map to a part of the data block.
//

export function get_object_type_id(object: QuestObject): number {
    return object.view.getUint16(0, true);
}

export function set_object_type_id(object: QuestObject, type_id: number): void {
    object.view.setUint16(0, type_id, true);
}

export function get_object_id(object: QuestObject): number {
    return object.view.getUint16(8, true);
}

export function get_object_group_id(object: QuestObject): number {
    return object.view.getUint16(10, true);
}

export function get_object_section_id(object: QuestObject): number {
    return object.view.getUint16(12, true);
}

export function set_object_section_id(object: QuestObject, section_id: number): void {
    object.view.setUint16(12, section_id, true);
}

/**
 * Section-relative position.
 */
export function get_object_position(object: QuestObject): Vec3 {
    return {
        x: object.view.getFloat32(16, true),
        y: object.view.getFloat32(20, true),
        z: object.view.getFloat32(24, true),
    };
}

export function set_object_position(object: QuestObject, position: Vec3): void {
    object.view.setFloat32(16, position.x, true);
    object.view.setFloat32(20, position.y, true);
    object.view.setFloat32(24, position.z, true);
}

export function get_object_rotation(object: QuestObject): Vec3 {
    return {
        x: angle_to_rad(object.view.getInt32(28, true)),
        y: angle_to_rad(object.view.getInt32(32, true)),
        z: angle_to_rad(object.view.getInt32(36, true)),
    };
}

export function set_object_rotation(object: QuestObject, rotation: Vec3): void {
    object.view.setInt32(28, rad_to_angle(rotation.x), true);
    object.view.setInt32(32, rad_to_angle(rotation.y), true);
    object.view.setInt32(36, rad_to_angle(rotation.z), true);
}

//
// Complex properties that use multiple parts of the data block and possibly other properties.
//

export function get_object_type(object: QuestObject): ObjectType {
    return id_to_object_type(get_object_type_id(object));
}

export function set_object_type(object: QuestObject, type: ObjectType): void {
    set_object_type_id(object, object_data(type).type_id ?? 0);
}

export function get_object_script_label(object: QuestObject): number | undefined {
    switch (get_object_type(object)) {
        case ObjectType.ScriptCollision:
        case ObjectType.ForestConsole:
        case ObjectType.TalkLinkToSupport:
            return object.view.getUint32(52, true);

        case ObjectType.RicoMessagePod:
            return object.view.getUint32(56, true);

        default:
            return undefined;
    }
}

export function get_object_script_label_2(object: QuestObject): number | undefined {
    switch (get_object_type(object)) {
        case ObjectType.RicoMessagePod:
            return object.view.getUint32(60, true);

        default:
            return undefined;
    }
}

export function get_object_model(object: QuestObject): number | undefined {
    switch (get_object_type(object)) {
        case ObjectType.Probe:
            return Math.round(object.view.getFloat32(40, true));

        case ObjectType.Saw:
        case ObjectType.LaserDetect:
            return Math.round(object.view.getFloat32(48, true));

        case ObjectType.Sonic:
        case ObjectType.LittleCryotube:
        case ObjectType.Cactus:
        case ObjectType.BigBrownRock:
        case ObjectType.BigBlackRocks:
        case ObjectType.BeeHive:
            return object.view.getUint32(52, true);

        case ObjectType.ForestConsole:
            return object.view.getUint32(56, true);

        case ObjectType.PrincipalWarp:
        case ObjectType.LaserFence:
        case ObjectType.LaserSquareFence:
        case ObjectType.LaserFenceEx:
        case ObjectType.LaserSquareFenceEx:
            return object.view.getUint32(60, true);

        default:
            return undefined;
    }
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/fileFormats/quest/QuestObject.kt
package world.phantasmal.lib.fileFormats.quest

import world.phantasmal.lib.buffer.Buffer
import world.phantasmal.lib.fileFormats.Vec3
import world.phantasmal.lib.fileFormats.ninja.angleToRad
import world.phantasmal.lib.fileFormats.ninja.radToAngle
import kotlin.math.roundToInt

class QuestObject(override var areaId: Int, override val data: Buffer) : QuestEntity<ObjectType> {
    constructor(type: ObjectType, areaId: Int) : this(areaId, Buffer.withSize(OBJECT_BYTE_SIZE)) {
        setObjectDefaultData(type, data)
        this.type = type
    }

    var typeId: Short
        get() = data.getShort(0)
        set(value) {
            data.setShort(0, value)
        }

    override var type: ObjectType
        get() = objectTypeFromId(typeId)
        set(value) {
            typeId = value.typeId ?: -1
        }

    var id: Short
        get() = data.getShort(8)
        set(value) {
            data.setShort(8, value)
        }

    var groupId: Short
        get() = data.getShort(10)
        set(value) {
            data.setShort(10, value)
        }

    override var sectionId: Short
        get() = data.getShort(12)
        set(value) {
            data.setShort(12, value)
        }

    override var position: Vec3
        get() = Vec3(data.getFloat(16), data.getFloat(20), data.getFloat(24))
        set(value) {
            setPosition(value.x, value.y, value.z)
        }

    override var rotation: Vec3
        get() = Vec3(
            angleToRad(data.getInt(28)),
            angleToRad(data.getInt(32)),
            angleToRad(data.getInt(36)),
        )
        set(value) {
            setRotation(value.x, value.y, value.z)
        }

    val scriptLabel: Int?
        get() = when (type) {
            ObjectType.ScriptCollision,
            ObjectType.ForestConsole,
            ObjectType.TalkLinkToSupport,
            -> data.getInt(52)

            ObjectType.RicoMessagePod,
            -> data.getInt(56)

            else -> null
        }

    val scriptLabel2: Int?
        get() = if (type == ObjectType.RicoMessagePod) data.getInt(60) else null

    /**
     * The offset of the model property or -1 if this object doesn't have a model property.
     */
    val modelOffset: Int
        get() = when (type) {
            ObjectType.Probe,
            -> 40

            ObjectType.Saw,
            ObjectType.LaserDetect,
            -> 48

            ObjectType.Sonic,
            ObjectType.LittleCryotube,
            ObjectType.Cactus,
            ObjectType.BigBrownRock,
            ObjectType.BigBlackRocks,
            ObjectType.BeeHive,
            -> 52

            ObjectType.ForestConsole,
            -> 56

            ObjectType.PrincipalWarp,
            ObjectType.LaserFence,
            ObjectType.LaserSquareFence,
            ObjectType.LaserFenceEx,
            ObjectType.LaserSquareFenceEx,
            -> 60

            else -> -1
        }

    var model: Int
        get() = when (type) {
            ObjectType.Probe,
            -> data.getFloat(40).roundToInt()

            ObjectType.Saw,
            ObjectType.LaserDetect,
            -> data.getFloat(48).roundToInt()

            ObjectType.Sonic,
            ObjectType.LittleCryotube,
            ObjectType.Cactus,
            ObjectType.BigBrownRock,
            ObjectType.BigBlackRocks,
            ObjectType.BeeHive,
            -> data.getInt(52)

            ObjectType.ForestConsole,
            -> data.getInt(56)

            ObjectType.PrincipalWarp,
            ObjectType.LaserFence,
            ObjectType.LaserSquareFence,
            ObjectType.LaserFenceEx,
            ObjectType.LaserSquareFenceEx,
            -> data.getInt(60)

            else -> throw IllegalArgumentException("$type doesn't have a model property.")
        }
        set(value) {
            when (type) {
                ObjectType.Probe,
                -> data.setFloat(40, value.toFloat())

                ObjectType.Saw,
                ObjectType.LaserDetect,
                -> data.setFloat(48, value.toFloat())

                ObjectType.Sonic,
                ObjectType.LittleCryotube,
                ObjectType.Cactus,
                ObjectType.BigBrownRock,
                ObjectType.BigBlackRocks,
                ObjectType.BeeHive,
                -> data.setInt(52, value)

                ObjectType.ForestConsole,
                -> data.setInt(56, value)

                ObjectType.PrincipalWarp,
                ObjectType.LaserFence,
                ObjectType.LaserSquareFence,
                ObjectType.LaserFenceEx,
                ObjectType.LaserSquareFenceEx,
                -> data.setInt(60, value)

                else -> throw IllegalArgumentException("$type doesn't have a model property.")
            }
        }

    val destinationPositionOffset: Int
        get() = when (type) {
            ObjectType.Warp, ObjectType.PrincipalWarp, ObjectType.RuinsWarpSiteToSite -> 40
            else -> -1
        }

    /**
     * Only valid for [ObjectType.Warp], [ObjectType.PrincipalWarp] and
     * [ObjectType.RuinsWarpSiteToSite].
     */
    var destinationPosition: Vec3
        get() = Vec3(
            data.getFloat(40),
            data.getFloat(44),
            data.getFloat(48),
        )
        set(value) {
            setDestinationPosition(value.x, value.y, value.z)
        }

    /**
     * Only valid for [ObjectType.Warp], [ObjectType.PrincipalWarp] and
     * [ObjectType.RuinsWarpSiteToSite].
     */
    var destinationPositionX: Float
        get() = data.getFloat(40)
        set(value) {
            data.setFloat(40, value)
        }

    /**
     * Only valid for [ObjectType.Warp], [ObjectType.PrincipalWarp] and
     * [ObjectType.RuinsWarpSiteToSite].
     */
    var destinationPositionY: Float
        get() = data.getFloat(44)
        set(value) {
            data.setFloat(44, value)
        }

    /**
     * Only valid for [ObjectType.Warp], [ObjectType.PrincipalWarp] and
     * [ObjectType.RuinsWarpSiteToSite].
     */
    var destinationPositionZ: Float
        get() = data.getFloat(48)
        set(value) {
            data.setFloat(48, value)
        }

    val destinationRotationYOffset: Int
        get() = when (type) {
            ObjectType.Warp, ObjectType.PrincipalWarp, ObjectType.RuinsWarpSiteToSite -> 52
            else -> -1
        }

    /**
     * Only valid for [ObjectType.Warp], [ObjectType.PrincipalWarp] and
     * [ObjectType.RuinsWarpSiteToSite].
     */
    var destinationRotationY: Float
        get() = angleToRad(data.getInt(52))
        set(value) {
            data.setInt(52, radToAngle(value))
        }

    init {
        require(data.size == OBJECT_BYTE_SIZE) {
            "Data size should be $OBJECT_BYTE_SIZE but was ${data.size}."
        }
    }

    override fun setPosition(x: Float, y: Float, z: Float) {
        data.setFloat(16, x)
        data.setFloat(20, y)
        data.setFloat(24, z)
    }

    override fun setRotation(x: Float, y: Float, z: Float) {
        data.setInt(28, radToAngle(x))
        data.setInt(32, radToAngle(y))
        data.setInt(36, radToAngle(z))
    }

    fun setDestinationPosition(x: Float, y: Float, z: Float) {
        data.setFloat(40, x)
        data.setFloat(44, y)
        data.setFloat(48, z)
    }
}
